File: src/commands/configure.ts
import type { Command } from "commander";
import { storeCredentials, getStoredUsername, getStoredToken, deleteCredentials } from "../keychain.ts";
import { normalizeUrl } from "../paths.ts";

export function registerConfigureCommands(program: Command): void {
  const configure = program
    .command("configure")
    .description("Manage stored Jenkins credentials");

  configure
    .command("set")
    .description("Store credentials for a Jenkins server")
    .requiredOption("-s <url>", "Jenkins server URL")
    .requiredOption("-u, --username <username>", "Jenkins username")
    .requiredOption("-t, --token <token>", "Jenkins API token")
    .action(async (opts: { s: string; username: string; token: string }) => {
      await storeCredentials(opts.s, opts.username, opts.token);
      console.log(`Credentials stored for ${normalizeUrl(opts.s)}`);
    });

  configure
    .command("show")
    .description("Show stored credentials for a Jenkins server")
    .requiredOption("-s <url>", "Jenkins server URL")
    .action(async (opts: { s: string }) => {
      const url = normalizeUrl(opts.s);
      const username = await getStoredUsername(url);
      if (!username) {
        console.log(`No credentials stored for ${url}`);
        return;
      }
      const token = await getStoredToken(url, username);
      console.log(`Server:   ${url}`);
      console.log(`Username: ${username}`);
      console.log(`Token:    ${token ? "********" : "(not found in Keychain)"}`);
    });

  configure
    .command("clear")
    .description("Remove stored credentials for a Jenkins server")
    .requiredOption("-s <url>", "Jenkins server URL")
    .action(async (opts: { s: string }) => {
      const url = normalizeUrl(opts.s);
      const removed = await deleteCredentials(url);
      if (removed) {
        console.log(`Credentials removed for ${url}`);
      } else {
        console.log(`No credentials found for ${url}`);
      }
    });
}

File: cmd/configure.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"github.com/nwp/jenkins-cli/internal/keyring"
	"github.com/nwp/jenkins-cli/internal/output"
	"github.com/nwp/jenkins-cli/internal/paths"
)

func newConfigureCmd() *cobra.Command {
	cfg := &cobra.Command{
		Use:   "configure",
		Short: "Manage stored Jenkins credentials",
	}
	cfg.AddCommand(newConfigureSetCmd(), newConfigureShowCmd(), newConfigureClearCmd())
	return cfg
}

func newConfigureSetCmd() *cobra.Command {
	var serverURL, username, token string
	cmd := &cobra.Command{
		Use:   "set",
		Short: "Store credentials for a Jenkins server",
		RunE: func(cmd *cobra.Command, args []string) error {
			normalized, err := paths.NormalizeURL(serverURL)
			if err != nil {
				return err
			}
			if err := keyring.StoreCredentials(normalized, username, token); err != nil {
				return fmt.Errorf("store credentials: %w", err)
			}
			fmt.Printf("Credentials stored for %s\n", normalized)
			return nil
		},
	}
	cmd.Flags().StringVarP(&serverURL, "server", "s", "", "Jenkins server URL")
	cmd.Flags().StringVarP(&username, "username", "u", "", "Jenkins username")
	cmd.Flags().StringVarP(&token, "token", "t", "", "Jenkins API token")
	cmd.MarkFlagRequired("server")
	cmd.MarkFlagRequired("username")
	cmd.MarkFlagRequired("token")
	return cmd
}

func newConfigureShowCmd() *cobra.Command {
	var serverURL string
	cmd := &cobra.Command{
		Use:   "show",
		Short: "Show stored credentials for a Jenkins server",
		RunE: func(cmd *cobra.Command, args []string) error {
			normalized, err := paths.NormalizeURL(serverURL)
			if err != nil {
				return err
			}
			username, err := keyring.GetStoredUsername(normalized)
			if err != nil {
				return err
			}
			if username == "" {
				fmt.Println("No credentials stored for that server.")
				return nil
			}
			token, err := keyring.GetStoredToken(normalized, username)
			if err != nil {
				return err
			}
			tokenDisplay := "********"
			if token == "" {
				tokenDisplay = "(not found in keyring)"
			}
			type credDisplay struct {
				Username string `json:"username"`
				Token    string `json:"token"`
			}
			if flagJSON {
				return output.Print(credDisplay{Username: username, Token: tokenDisplay}, true)
			}
			fmt.Printf("Username: %s\n", username)
			fmt.Printf("Token:    %s\n", tokenDisplay)
			return nil
		},
	}
	cmd.Flags().StringVarP(&serverURL, "server", "s", "", "Jenkins server URL")
	cmd.MarkFlagRequired("server")
	return cmd
}

func newConfigureClearCmd() *cobra.Command {
	var serverURL string
	cmd := &cobra.Command{
		Use:   "clear",
		Short: "Remove stored credentials for a Jenkins server",
		RunE: func(cmd *cobra.Command, args []string) error {
			normalized, err := paths.NormalizeURL(serverURL)
			if err != nil {
				return err
			}
			if err := keyring.DeleteCredentials(normalized); err != nil {
				return fmt.Errorf("delete credentials: %w", err)
			}
			fmt.Printf("Credentials cleared for %s\n", normalized)
			return nil
		},
	}
	cmd.Flags().StringVarP(&serverURL, "server", "s", "", "Jenkins server URL")
	cmd.MarkFlagRequired("server")
	return cmd
}
